src/provider/mod.rs

```diff
@@ -243,6 +261,244 @@ where
+    /// This constant is part of a coefficient calculation to adjust the gas limit to account for variations
+    /// in the SDK estimation, ensuring the transaction will be accepted.
+    const L1_FEE_ESTIMATION_COEF_DENOMINATOR: u64 = 10;
+    l1_gas_limit * L1_FEE_ESTIMATION_COEF_NUMERATOR / L1_FEE_ESTIMATION_COEF_DENOMINATOR
+}
+
+/// Enum to describe errors that might occur during L1 -> L2 communication.
+#[derive(Debug, thiserror::Error)]
+pub enum L1CommunicationError {
+    #[error("NewPriorityRequest event log was not found in L1 -> L2 transaction.")]
+    NewPriorityRequestLogNotFound,
+    #[error("Custom L1 -> L2 communication error.")]
+    Custom(&'static str),
+}
+
+/// A wrapper struct to hold L1 transaction receipt and L2 provider
+/// which is used by the associated functions.
+pub struct L1TransactionReceipt<T> {
+    /// Ethereum transaction receipt.
+    inner: TransactionReceipt,
+    /// A reference to the L2 provider.
+    l2_provider: RootProvider<T, Zksync>,
+}
+
+impl<T> L1TransactionReceipt<T>
+where
+    T: Transport + Clone,
+{
+    pub fn new(tx_receipt: TransactionReceipt, l2_provider: RootProvider<T, Zksync>) -> Self {
+        Self {
+            inner: tx_receipt,
+            l2_provider,
+        }
+    }
+
+    pub fn get_receipt(&self) -> &TransactionReceipt {
+        &self.inner
+    }
+
+    pub fn get_l2_tx(&self) -> Result<PendingTransactionBuilder<T, Zksync>, L1CommunicationError> {
+        let l1_to_l2_tx_log = self
+            .inner
+            .inner
+            .logs()
+            .iter()
+            .filter_map(|log| log.log_decode::<Bridgehub::NewPriorityRequest>().ok())
+            .next()
+            .ok_or(L1CommunicationError::NewPriorityRequestLogNotFound)?;
+
+        let l2_tx_hash = l1_to_l2_tx_log.inner.txHash;
+
+        Ok(PendingTransactionBuilder::new(
+            self.l2_provider.clone(),
+            l2_tx_hash,
+        ))
+    }
+}
+
+/// Trait for ZKsync provider with populated wallet
+/// Contains provider methods that need a wallet
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+pub trait ZksyncProviderWithWallet<T = BoxTransport>:
+    ZksyncProvider<T> + WalletProvider<Zksync>
+where
+    T: Transport + Clone,
+{
+    /// Deposits specified L1 token to the L2 address.
+    ///
+    /// ## Parameters
+    ///
+    /// - `token_address`: L1 token address to deposit.
+    /// - `receiver`: receiver L2 address.
+    /// - `amount`: amount to deposit in Wei.
+    /// - `l1_provider`: reference to the L1 provider.
+    ///
+    /// ## Returns
+    ///
+    /// L1TransactionReceipt.
+    /// Hint: use returned L1 transaction receipt to get corresponding L2 transaction and wait for its receipt
+    /// E.g.: deposit_l1_receipt.get_l2_tx()?.with_required_confirmations(1).with_timeout(Some(std::time::Duration::from_secs(60 * 5))).get_receipt()
```

Comment:
Nit: it's better to structure it as a code block with `ignore` attribute. Actually, ideally you should write the snipped with hidden parts of the code so that it at least compiles, but we can skip that for now :)

See [rustdoc docs](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html) for reference.